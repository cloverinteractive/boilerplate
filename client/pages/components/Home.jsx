// @flow

import React from 'react';
import { connect } from 'react-redux';
import { Button, Container, Divider, Header, Segment } from 'semantic-ui-react';
import { Helmet } from 'react-helmet';
import { MAIN_TREE, PAGES_TREE } from 'pages/constants/home';
import * as messages from 'main/flux/actions/messages';
import 'pages/styles/code.css';

type Props = {
  addError: Function,
  addNotice: Function,
  addWarning: Function,
};

const Home = (props: Props) => (
  <Container>
    <Helmet>
      <title>Home</title>
    </Helmet>

    <Header as="h1">Everything is Running!</Header>

    <p>Awesome, you&apos;ve got this far, here are some pointers to make development easier.</p>

    <Divider section />

    <Header as="h2">Bundled goodies</Header>

    <p>Here&apos;s the list of the the packages this boiler plate configures for you.</p>

    <ul>
      <li>
        <a href="https://babeljs.io">Babel</a> - <em>
          The compiler for writing next generation JavaScript
        </em>
      </li>
      <li>
        <a href="https://eslint.org">Eslint</a> - <em>
          Pluggable JavaScript linter
        </em>
      </li>
      <li>
        <a href="https://expressjs.com">Express</a> - <em>
          Fast, unopinionated, minimalist web framework for Node.js
        </em>
      </li>
      <li>
        <a href="https://flow.org">Flow</a> - <em>
          A Static Type Checker for JavaScript
        </em>
      </li>
      <li>
        <a href="https://mochajs.org">Mocha</a> - <em>
          The fun, simple, flexible JavaScript test framework
        </em>
      </li>
      <li>
        <a href="https://facebook.github.io/react/">React</a> - <em>
          A JavaScript library for building user interfaces
        </em>
      </li>
      <li>
        <a href="https://reacttraining.com/react-router/">React Router</a> - <em>
          Declarative Routing for React.js
        </em>
      </li>
      <li>
        <a href="http://redux.js.org/">Redux</a> - <em>
          Redux is a predictable state container for JavaScript apps.
        </em>
      </li>
      <li>
        <a href="https://semantic-ui.com">SemanticUI</a> - <em>
          Semantic is a development framework that helps create beautiful, responsive
          layouts using human-friendly HTML.
        </em>
      </li>
      <li>
        <a href="https://webpack.js.org/concepts/">Webpack</a> - <em>
          Webpack is a module bundler for modern JavaScript applications.
        </em>
      </li>
    </ul>

    <Divider section />

    <Header as="h2">Components</Header>

    <p>
      The app comes bundled with a super basic <strong><em>redux</em></strong> reducer for adding
      and removing <strong><em>SemanticUI</em></strong>
      <a
        href="https://react.semantic-ui.com/collections/message"
        rel="noopener noreferrer"
        target="_blank"
      >
        &lt;Message /&gt;
      </a>
      . You can find the code for
      these, under the <strong><em>client/main</em></strong> folder. You can also try
      the basic messaging by clicking any of the following buttons.
    </p>

    <Button.Group>
      <Button color="green" onClick={() => { props.addNotice('This is a successful message'); }}>
        Notice
      </Button>
      <Button.Or />
      <Button color="yellow" onClick={() => { props.addWarning('This is a warning message'); }}>
        Warning
      </Button>
      <Button.Or />
      <Button color="red" onClick={() => { props.addError('This is an error message'); }}>
        Error
      </Button>
    </Button.Group>

    <p>Folder structure looks like this: </p>

    <Segment className="code" inverted>
      <code>
        <pre>{ MAIN_TREE }</pre>
      </code>
    </Segment>

    <p>
      It also comes pre-packaged with 3 pages, <strong><em>&lt;Home /&gt;</em></strong>,
      <strong><em> &lt;About /&gt;</em></strong> and <strong><em>&lt;Error404 /&gt;</em></strong>.
      You can use the navigation at the top to visit the first two, and type any unexisting
      page in the browser URL to visit the latter.
    </p>

    <p>
      We&apos;ve organized the application by features and created a new folder per each new
      feature, the aforementions components can be found at <em>client/pages</em> in this app&apos;s
      folder. Folder structure looks like this:
    </p>

    <Segment className="code" inverted>
      <code>
        <pre>{ PAGES_TREE }</pre>
      </code>
    </Segment>


    <p>
      Check our <strong>
        <a href="https://github.com/cloverinteractive/boilerplate">
          README
        </a>
      </strong> for more details.
    </p>
  </Container>
);

export default connect(null, messages)(Home);
