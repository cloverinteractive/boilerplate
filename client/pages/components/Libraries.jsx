// @flow

import React from 'react';
import { v4 } from 'uuid';
import libraries from 'pages/constants/libraries';

type NpmLibrary = {
  description: string,
  library: string,
  url: string,
};

const Library = ({ description, library, url }: NpmLibrary) => (
  <li>
    <a href={url} rel="noopener noreferrer" target="_blank">{library}</a> - <em>{description}</em>
  </li>
);

const Libraries = () => {
  const items = libraries.map(library => <Library key={v4()} {...library} />);

  return (
    <ul className="libraries">
      { items }
    </ul>
  );
};

export default Libraries;
