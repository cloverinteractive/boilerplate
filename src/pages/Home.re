open HomePageMocks;

[@react.component]
let make = () =>
  MessageContext.(
    {
      let (_, dispatch) = useMessages();
      let highlightClass =
        Cn.make(["has-text-weight-bold", "is-family-monospace", "is-italic"]);

      <div className="container is-fluid">
        <article className="content">
          <h1 className="title"> "Everything is Running!"->React.string </h1>
          <p>
            "Awesome, you've got this far, here are some pointers to make development easier."
            ->React.string
          </p>
        </article>
        <article className="content">
          <h2 className="title"> "Bundled goodies"->React.string </h2>
          <p>
            "Here's the list of packages this boilerplate pre-configures for you."
            ->React.string
          </p>
          <ul className="libraries">
            {librariesList
             ->Belt.List.map(
                 ({description, library, url}: HomePageMocks.library) => {
                 <li key=library>
                   <a href=url rel="noopener noreferrer" target="_blank">
                     library->React.string
                   </a>
                   " - "->React.string
                   <span className="is-italic">
                     description->React.string
                   </span>
                 </li>
               })
             ->Array.of_list
             ->React.array}
          </ul>
        </article>
        <article className="content">
          <h2 className="title"> "Components"->React.string </h2>
          <p>
            "The app comes bundled with super basic notification and dismissable components
        you can try them out by clicking the buttons below."
            ->React.string
          </p>
          <div className="field is-grouped is-grouped-centered">
            <div className="control">
              <button
                className="button is-success"
                onClick={_ => dispatch(AddMessage("is-success", "Success!"))}
                title="Successful alerts will automatically close"
                type_="button">
                "Notice"->React.string
              </button>
            </div>
            <div className="control">
              <button
                className="button is-warning"
                onClick={_ => dispatch(AddMessage("is-warning", "Warning!"))}
                title="Warning alerts will stay until manually closed"
                type_="button">
                "Warning"->React.string
              </button>
            </div>
            <div className="control">
              <button
                className="button is-danger"
                onClick={_ => dispatch(AddMessage("is-danger", "Error!"))}
                title="Error alerts will stay until manually closed"
                type_="button">
                "Error"->React.string
              </button>
            </div>
          </div>
          <p>
            "We've also packaged 3 page components,"->React.string
            <span className=highlightClass> "<Home />"->React.string </span>
            ", "->React.string
            <span className=highlightClass>
              "<Error500 />"->React.string
            </span>
            " and "->React.string
            <span className=highlightClass>
              "<Error404 />"->React.string
            </span>
            " and organized the application in a simple structure making it easy to find everyting in the project;"
            ->React.string
          </p>
          <p>
            "This is the "->React.string
            <span className=highlightClass> "src/pages"->React.string </span>
            " structure:"->React.string
          </p>
          <pre className="is-family-monospace"> pagesTree->React.string </pre>
          <p>
            "This is the "->React.string
            <span className=highlightClass>
              "src/components"->React.string
            </span>
          </p>
          <pre className="is-family-monospace">
            componentTree->React.string
          </pre>
        </article>
        <BsReactHelmet defaultTitle="Home Page" />
      </div>;
    }
  );

let default = make;