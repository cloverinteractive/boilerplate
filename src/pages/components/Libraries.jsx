import React from 'react';
import v4 from 'uuid/v4';

export const Library = ({ description, library, url }) => (
  <li>
    <a href={url} rel="noopener noreferrer" target="_blank">{library}</a> - <em>{description}</em>
  </li>
);

const Libraries = ({ libraries }) => {
  const items = libraries.map(library => <Library key={v4()} {...library} />);

  return (
    <ul className="libraries">
      { items }
    </ul>
  );
};

export default Libraries;
