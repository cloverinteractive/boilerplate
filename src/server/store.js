import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import rootReducer from 'root-reducer';

const defaultState = {};
const enhancers = applyMiddleware(thunkMiddleware);
const store = createStore(rootReducer, defaultState, enhancers);

export default store;
