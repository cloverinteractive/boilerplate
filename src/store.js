/* eslint-disable no-underscore-dangle */

import { createStore, compose, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import rootReducer from './root-reducer';

export const history = createBrowserHistory();

const isDevelop = process.env.NODE_ENV !== 'production';
const initialState = window.__INITIAL_STATE__;
const composeEnhancer = (isDevelop && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

const store = createStore(
  connectRouter(history)(rootReducer),
  initialState,
  composeEnhancer(applyMiddleware(
    routerMiddleware(history),
    thunkMiddleware,
  )),
);

export default store;
