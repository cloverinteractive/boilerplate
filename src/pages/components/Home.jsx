import React from 'react';
import { connect } from 'react-redux';
import { Button, Container, Header, Popup, Segment } from 'semantic-ui-react';
import { Helmet } from 'react-helmet';
import { MAIN_TREE, PAGES_TREE } from 'pages/constants/home';
import * as messages from 'main/flux/actions/messages';

import libraries from 'pages/constants/libraries';
import Libraries from 'pages/components/Libraries';

import styles from 'pages/css/code.css';

const Home = ({ addError, addNotice, addWarning }) => {
  const errorButton = (
    <Button color="red" onClick={() => { addError('This is an error message'); }}>
      Error
    </Button>
  );

  const noticeButton = (
    <Button color="green" onClick={() => { addNotice('This is a successful message'); }}>
      Notice
    </Button>
  );

  const warningButton = (
    <Button color="yellow" onClick={() => { addWarning('This is a warning message'); }}>
      Warning
    </Button>
  );

  return (
    <Container>
      <Helmet>
        <title>Home</title>
      </Helmet>

      <Segment vertical>
        <Header as="h1">Everything is Running!</Header>
        <p>Awesome, you&apos;ve got this far, here are some pointers to make development easier.</p>
      </Segment>

      <Segment vertical>
        <Header as="h2">Bundled goodies</Header>
        <p>Here&apos;s the list of the the packages this boiler plate configures for you.</p>
        <Libraries libraries={libraries} />
      </Segment>

      <Segment vertical>
        <Header as="h2">Components</Header>
        <p>
          The app comes bundled with a super basic <strong><em>redux</em></strong> reducer for
          adding and removing <strong><em>SemanticUI</em></strong>
          <a
            href="https://react.semantic-ui.com/collections/message"
            rel="noopener noreferrer"
            target="_blank"
          >
            &lt;Message /&gt;
          </a>
          . You can find the code for
          these, under the <strong><em>client/main</em></strong> folder. You can also try
          the basic messaging by clicking any of the following buttons.
        </p>

        <Button.Group>
          <Popup
            content="Successful alerts will automatically close"
            inverted
            size="mini"
            trigger={noticeButton}
          />
          <Button.Or />
          <Popup
            content="Warning alerts will hang until manually closed"
            inverted
            size="mini"
            trigger={warningButton}
          />
          <Button.Or />
          <Popup
            content="Error alerts will hang until manually closed"
            inverted
            size="mini"
            trigger={errorButton}
          />
        </Button.Group>

        <p>Folder structure looks like this: </p>

        <Segment className={styles.codeWrapper} inverted>
          <code>
            <pre className={styles.code}>{MAIN_TREE}</pre>
          </code>
        </Segment>

        <p>
          It also comes pre-packaged with 3 pages, <strong><em>&lt;Home /&gt;</em></strong>,
          <strong><em> &lt;About /&gt;</em></strong> and
          <strong><em>&lt;Error404 /&gt;</em></strong>. You can use the navigation at the top
          to visit the first two, and type any unexisting page in the browser URL to visit the
          latter.
        </p>

        <p>
          We&apos;ve organized the application by features and created a new folder per each new
          feature, the aforementions components can be found at <em>client/pages</em> in this
          app&apos;s folder. Folder structure looks like this:
        </p>

        <Segment className={styles.codeWrapper} inverted>
          <code>
            <pre className={styles.code}>{PAGES_TREE}</pre>
          </code>
        </Segment>

      </Segment>

      <Segment vertical>
        <p>
          Check our
          { ' ' }
          <strong>
            <a href="https://github.com/cloverinteractive/boilerplate">
              README
            </a>
          </strong>
          { ' ' }
          for more details.
        </p>
      </Segment>
    </Container>
  );
};

export default connect(null, messages)(Home);
