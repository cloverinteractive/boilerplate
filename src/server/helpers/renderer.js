import { renderToString } from 'react-dom/server';
import { Helmet } from 'react-helmet';
import serialize from 'serialize-javascript';
import { isDevelopment } from 'server/helpers/env-access';

export default (Component, store) => {
  const content = renderToString(Component);
  const helmet = Helmet.renderStatic();

  const productionStyles =
    `<link rel="stylesheet" href="/vendors.css">
    <link rel="stylesheet" href="/styles.css">`;

  const styles = isDevelopment ? '' : productionStyles;

  const template =
    `<!doctype html>
    <html lang="en-US">
      <head>
        <meta charset="utf-8">
        ${helmet.link.toString()}
        ${helmet.title.toString()}
        ${helmet.meta.toString()}
        ${styles}
        <script src="/vendors.js"></script>
      </head>
      <body>
        <div id="app">${content}</div>
        <script>
          window.__INITIAL_STATE__ = ${serialize(store.getState())};
        </script>
        <script async src="/bundle.js"></script>
      </body>
    </html>`;

  return template;
};
