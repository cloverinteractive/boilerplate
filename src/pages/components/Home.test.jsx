import * as React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { act } from '@testing-library/react';
import { mount } from 'enzyme';
import Home from './Home';
import Libraries from './Libraries';
import store from '../../redux/store';

jest.mock('uuid/v1', () => jest.fn((_, increment) => `mocked-random-uuid-${increment}`));

describe('<Home />', () => {
  test('it renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Home />
        </Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('it renders libraries', () => {
    const wrapper = mount(
      <Provider store={store}>
        <Home />
      </Provider>,
    );

    expect(wrapper.find(Libraries)).toHaveLength(1);
  });

  test('it clicks buttons', () => {
    const wrapper = mount(
      <Provider store={store}>
        <Home />
      </Provider>,
    );

    act(() => {
      wrapper.find('button.is-success').simulate('click');
      wrapper.find('button.is-warning').simulate('click');
      wrapper.find('button.is-danger').simulate('click');
    });
  });
});
