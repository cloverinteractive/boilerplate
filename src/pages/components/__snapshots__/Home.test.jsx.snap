// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> it renders correctly 1`] = `
<div
  className="container is-fluid"
>
  <article
    className="content"
  >
    <h1
      className="title"
    >
      Everything is Running!
    </h1>
    <p>
      Awesome, you've got this far, here are some pointers to make development easier.
    </p>
  </article>
  <article
    className="content"
  >
    <h2
      className="title"
    >
      Bundled goodies
    </h2>
    <p>
      Here's the list of the the packages this boiler plate configures for you.
    </p>
    <ul
      className="libraries"
    >
      <li>
        <a
          href="https://babeljs.io"
          rel="noopener noreferrer"
          target="_blank"
        >
          Babel
        </a>
         
        -
        <span
          className="is-italic"
        >
          The compiler for writing next generation JavaScript
        </span>
      </li>
      <li>
        <a
          href="https://eslint.org"
          rel="noopener noreferrer"
          target="_blank"
        >
          Eslint
        </a>
         
        -
        <span
          className="is-italic"
        >
          Pluggable JavaScript linter
        </span>
      </li>
      <li>
        <a
          href="https://expressjs.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          Express
        </a>
         
        -
        <span
          className="is-italic"
        >
          Fast, unopinionated, minimalist web framework for Node.js
        </span>
      </li>
      <li>
        <a
          href="https://reasonml.github.io/"
          rel="noopener noreferrer"
          target="_blank"
        >
          ReasonML
        </a>
         
        -
        <span
          className="is-italic"
        >
          Reason lets you write simple, fast and quality type safe code while leveraging both the JavaScript & OCaml ecosystems.
        </span>
      </li>
      <li>
        <a
          href="https://www.typescriptlang.org/"
          rel="noopener noreferrer"
          target="_blank"
        >
          TypeScript
        </a>
         
        -
        <span
          className="is-italic"
        >
          A typed superset of JavaScript that compiles to plain Javascript
        </span>
      </li>
      <li>
        <a
          href="https://facebook.github.io/react/"
          rel="noopener noreferrer"
          target="_blank"
        >
          React
        </a>
         
        -
        <span
          className="is-italic"
        >
          A JavaScript library for building user interfaces
        </span>
      </li>
      <li>
        <a
          href="https://reacttraining.com/react-router/"
          rel="noopener noreferrer"
          target="_blank"
        >
          React Router
        </a>
         
        -
        <span
          className="is-italic"
        >
          Declarative Routing for React.js
        </span>
      </li>
      <li>
        <a
          href="http://redux.js.org/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Redux
        </a>
         
        -
        <span
          className="is-italic"
        >
          Redux is a predictable state container for JavaScript apps.
        </span>
      </li>
      <li>
        <a
          href="https://webpack.js.org/concepts/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Webpack
        </a>
         
        -
        <span
          className="is-italic"
        >
          Webpack is a module bundler for modern JavaScript applications.
        </span>
      </li>
    </ul>
  </article>
  <article
    className="content"
  >
    <h2
      className="title"
    >
      Components
    </h2>
    <p>
      The app comes bundled with a super basic
       
      <span
        className="has-text-weight-bold is-family-monospace is-italic"
      >
        redux
      </span>
       
      reducer for adding and removing alerts. You can find the code for these, under the
       
      <span
        className="has-text-weight-bold is-family-monospace is-italic"
      >
        src/redux
      </span>
       
      folder. You can also try the basic messaging by clicking any of the following buttons.
    </p>
    <div
      className="field is-grouped is-grouped-centered"
    >
      <div
        className="control"
      >
        <button
          className="button is-success"
          onClick={[Function]}
          title="Successful alerts will automatically close"
          type="button"
        >
          Notice
        </button>
      </div>
      <div
        className="control"
      >
        <button
          className="button is-warning"
          onClick={[Function]}
          title="Warning alerts will stay until manually closed"
          type="button"
        >
          Warning
        </button>
      </div>
      <div
        className="control"
      >
        <button
          className="button is-danger"
          onClick={[Function]}
          title="Error alerts will stay until manually closed"
          type="button"
        >
          Error
        </button>
      </div>
    </div>
    <p>
      Folder structure looks like this:
    </p>
    <pre
      className="is-family-monospace"
    >
      src/redux/
├── actions
│   ├── messages.js
│   └── types.js
├── reducers
│   ├── index.js
│   └── messages.js
├── selectors
│   └── messages.js
└── store.js
    </pre>
    <p>
      It also comes pre-packaged with 3 pages,
      <span
        className="has-text-weight-bold is-family-monospace is-italic"
      >
         &lt;Home /&gt;
      </span>
      ,
      <span
        className="has-text-weight-bold is-family-monospace is-italic"
      >
         &lt;About /&gt;
      </span>
       
      and
      <span
        className="has-text-weight-bold is-family-monospace is-italic"
      >
         &lt;Error404 /&gt;
      </span>
      . You can use the navigation at the top to visit the first two, and type any unexisting page in the browser URL to visit the latter.
    </p>
    <p>
      We've organized the application by features and created a new folder per each new feature, the aforementions components can be found at
       
      <span
        className="has-text-weight-bold is-family-monospace is-italic"
      >
        src/pages
      </span>
       
      in this app's folder. Folder structure looks like this:
    </p>
    <pre
      className="is-family-monospace"
    >
      src/pages/
├── components
│   ├── About.jsx
│   ├── Error404.jsx
│   ├── Home.jsx
│   └── Libraries.jsx
├── constants
│   ├── home.js
│   └── libraries.js
└── index.js
    </pre>
    <p>
      Lastly we've shipped a few regular purpose components, you can guess what they are
       
      for by looking at their name.
    </p>
    <p>
      Folder structure looks like this:
    </p>
    <pre
      className="is-family-monospace"
    >
      src/components/
├── Dismissable.re
├── Messages.jsx
├── Nav.re
├── Notification.re
└── ReactRouter.re
    </pre>
  </article>
</div>
`;
